### PR TITLE
[MAINTENANCE] Add type-hints; add utility method for creating temporary DB tables; clean up imports; improve code readability; and add a directory to pre-commit

### PR DESCRIPTION
**Pull request by alexsherstinsky**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: GREAT-56/GREAT-93/GREAT-295
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 1]
```diff
@@ -375,7 +374,7 @@ def _get_count_query_mssql(
         ignore_values_condition: BinaryExpression,
     ) -> Select:
         # mssql expects all temporary table names to have a prefix '#'
```
Comment (NathanFarmer) [inline great_expectations/dataset/sqlalchemy_dataset.py]: Is this comment still needed? Does this function handle it based on dialect?

File: great_expectations/util.py [hunk 2]
```diff
@@ -1069,3 +1070,11 @@ def generate_library_json_from_registered_expectations():
 
 def delete_blank_lines(text: str) -> str:
     return re.sub(r"\n\s*\n", "\n", text, flags=re.MULTILINE)
+
+
+def generate_temporary_table_name(
+    default_table_name_prefix: Optional[str] = "ge_tmp_",
+    num_digits: Optional[int] = 8,
+) -> str:
+    table_name: str = f"{default_table_name_prefix}{str(uuid.uuid4())[:num_digits]}"
```
Comment (NathanFarmer) [inline great_expectations/util.py:1079]: What about dialect-specific cases such as MSSQL requiring temp tables to start with `#`?

**Review by NathanFarmer — CHANGES_REQUESTED** [review]
Explanation for my 2 comments or updates to `generate_temporary_table_name` should get this all the way there.

File: great_expectations/util.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/util.py:1079]: @NathanFarmer One can override the default prefix.  Let me fix this.  For some reason "grep" missed this one.  Nice catch.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -19,7 +19,7 @@ repos:
     rev: 20.8b1
     hooks:
       - id: black
-        exclude: docs/.*|tests/.*.fixture|.*.ge_store_backend_id
+        exclude: docs/.*|tests/.*.fixture|.*.ge_store_backend_id|DO_NOT_COMMIT/.*
```
Comment (anthonyburdi) [inline .pre-commit-config.yaml]: Is this something that can be added to your local configuration instead of to the library?

Comment (alexsherstinsky) [reply .pre-commit-config.yaml]: deleted

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 4]
```diff
@@ -375,7 +374,7 @@ def _get_count_query_mssql(
...
-        temp_table_name: str = f"#ge_tmp_{str(uuid.uuid4())[:8]}"
+        temp_table_name: str = generate_temporary_table_name()
```
Comment (NathanFarmer) [inline great_expectations/dataset/sqlalchemy_dataset.py]: I believe this should be `generate_temporary_table_name(default_table_name_prefix="#ge_temp_")`. I wonder why this is passing tests?

**Review by NathanFarmer — CHANGES_REQUESTED** [review]
I believe you missed one more for mssql.

File: great_expectations/dataset/sqlalchemy_dataset.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/dataset/sqlalchemy_dataset.py]: @NathanFarmer Not sure, but I think that you are correct -- let me fix it.  Great catch.  Thank you!

Comment (alexsherstinsky) [reply great_expectations/dataset/sqlalchemy_dataset.py]: Done -- please check.  Thanks.

**Comment by alexsherstinsky** [comment]
> I believe you missed one more for mssql.

@NathanFarmer Fixed -- please verify.  Thanks!

File: great_expectations/dataset/sqlalchemy_dataset.py [hunk 5]
```diff
@@ -511,7 +513,9 @@ def __init__(
             # NOTE: Eugene 2020-01-31: @James, this is a not a proper fix, but without it the "public" schema
             # was used for a temp table and raising an error
             schema = None
-            table_name = generate_temporary_table_name()
+            table_name = generate_temporary_table_name(
+                default_table_name_prefix="#ge_temp_"
```
Comment (NathanFarmer) [inline great_expectations/dataset/sqlalchemy_dataset.py]: I don't believe `#` is needed here because it is added below if dialect is mssql.

**Review by NathanFarmer — CHANGES_REQUESTED** [review]
Ok, now I think you added one too many. 😂

File: great_expectations/dataset/sqlalchemy_dataset.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/dataset/sqlalchemy_dataset.py]: No -- deleted.  Great catch again (deleted from V3, but missed V2) -- please verify!  Thanks!

File: great_expectations/dataset/sqlalchemy_dataset.py (hunk 5 above)
Comment (alexsherstinsky) [reply great_expectations/dataset/sqlalchemy_dataset.py]: Fixed.

**Review by NathanFarmer — APPROVED** [review]
LGTM! 😄

**Comment by alexsherstinsky** [comment]
> LGTM! 😄

🙇 Thanks for the many great comments, @NathanFarmer !

**Pull request merged**